File: 1522-Diameter of N-Ary Tree.py
"""
# Definition for a Node.
class Node(object):
    def __init__(self, val=None, children=None):
        self.val = val
        self.children = children if children is not None else []
"""

class Solution(object):
    def diameter(self, root):
        """
        :type root: 'Node'
        :rtype: int
        """
        self.res = 0
        
        def dfs(root):
            if not root.children:
                return 0
            
            firstmax, secondmax = 0, 0
            for nxt in root.children:
                val = dfs(nxt)
                if val + 1 > firstmax:
                    secondmax = firstmax
                    firstmax = val + 1
                elif val + 1 > secondmax:
                    secondmax = val + 1
            self.res = max(self.res, firstmax + secondmax)
            return firstmax
        dfs(root)
        return self.res

File: JavaScript/1522-Diameter of N-Ary Tree.js
/**
 * // Definition for a Node.
 * function Node(val, children) {
 *    this.val = val === undefined ? 0 : val;
 *    this.children = children === undefined ? [] : children;
 * };
 */

/**
 * @param {Node} root
 * @return {number}
 */
var diameter = function(root) {
    let res = 0;
    
    let dfs = (root) =>{
        if (!root.children.length){
            return 0;
        }
        let [firstmax, secondmax] = [0, 0];
        for (let nxt of root.children){
            let value = dfs(nxt);
            if (value + 1 > firstmax){
                secondmax = firstmax
                firstmax = value + 1;
            } else if (value + 1 > secondmax){
                secondmax = value + 1;
            }
        }
        res = Math.max(res, firstmax + secondmax);
        return firstmax;
    }
    dfs(root);
    return res;
};